state-chain/ethereum-eip712/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+	comp_value: Composite<C>,
+	types: &mut BTreeMap<String, Vec<Eip712DomainType>>,
+) -> Result<Vec<Eip712DomainType>, &'static str> {
+	match comp_value {
+		Composite::Named(fs) => {
+			let fs_map = fs.into_iter().collect::<BTreeMap<_, _>>();
+			fields
+				.clone()
+				.into_iter()
+				.map(|field| -> Result<_, &'static str> {
+					// shouldn't be possible since we are in Named variant
+					let field_name = field.name.ok_or("field name doesn't exist")?.to_string();
+					let value =
+						fs_map.get(&field_name).ok_or("field with this name has to exist")?.clone();
+					Ok(Eip712DomainType {
+						name: field_name.clone(),
+						// find out in which cases type name would be empty.
+						r#type: recursively_construct_types(value.clone(), field.ty, types)?,
+					})
+				})
+				.collect::<Result<_, _>>()
+		},
+		Composite::Unnamed(fs) => {
+			fields
+				.clone()
+				.into_iter()
+				.zip(fs)
+				.enumerate()
+				.map(|(i, (field, value))| -> Result<_, &'static str> {
+					let type_name = recursively_construct_types(value.clone(), field.ty, types)?;
+					Ok(Eip712DomainType {
+						// In case of unnamed fields, we decide to name it by its type name appended
+						// by its index in the tuple
+						name: type_name.clone() + "_" + &i.to_string(),
+						r#type: type_name,
+					})
+				})
+				.collect::<Result<_, _>>()
+		},
+	}
+}
+
+impl<C: Clone> GetScaleValueFields for scale_value::Value<C> {
+	fn get_scale_value_fields(&self) -> Result<Vec<Value<C>>, &'static str> {
+		match &self.value {
+			ValueDef::Composite(comp) => Ok(match comp.clone() {
+				Composite::Named(fs) => fs.into_iter().map(|(_, v)| v).collect(),
+				Composite::Unnamed(fs) => fs,
+			}),
+			ValueDef::Variant(v) => Ok(match v.values.clone() {
+				Composite::Named(fs) => fs.into_iter().map(|(_, v)| v).collect(),
+				Composite::Unnamed(fs) => fs,
+			}),
+			ValueDef::Primitive(_) => Err("Primitive type does not have fields"),
+			ValueDef::BitSequence(_) => Err("BitSequence not supported"),
+		}
+	}
+}
+
+pub trait GetScaleValueFields: Sized {
+	fn get_scale_value_fields(&self) -> Result<Vec<Self>, &'static str>;
+}
+
+#[cfg(test)]
+pub mod tests {
+
+	use super::*;
+	use ethabi::ethereum_types::{H160, U256};
+	use scale_info::prelude::string::String;
+	//se serde::Deserialize;
+
+	#[test]
+	fn test_type_info_eip_712() {
+		let domain = EIP712Domain {
+			name: Some(String::from("Seaport")),
+			version: Some(String::from("1.1")),
+			chain_id: Some(U256([1, 0, 0, 0])),
+			verifying_contract: Some(H160([0xef; 20])),
+			salt: None,
+		};
+
+		#[derive(TypeInfo, Clone, Encode, Decode)]
+		pub struct Mail {
+			pub from: Person,
+			pub to: Person,
+			pub message: String,
+		}
+
+		#[derive(TypeInfo, Clone, Encode, Decode)]
+		pub struct Person {
+			pub name: String,
+		}
+
+		let payload = Mail {
+			from: Person { name: String::from("Ramiz") },
+			to: Person { name: String::from("Albert") },
+			message: String::from("hello Albert"),
```

Comment:
I doubt this is in the official test vectors 😅 
How did we generate the reference output? Would be good to document this, or otherwise to use an official reference value from the EIP.